### PR TITLE
Adapt K8s connection settings creation to newest schema version

### PR DESCRIPTION
**Pull request by annazaionchkovska**
# Description

After splitting k8s setting to Connection and Monitoring setting creation of k8s should be also adopted.
Description of split you can find at:  [Info](https://www.dynatrace.com/support/help/platform-modules/infrastructure-monitoring/container-platform-monitoring/kubernetes-monitoring/default-monitoring-settings)

## Approach 1 (removed after introducing approach 2)
Depending on schema "builtin:cloud.kubernetes" version we use  "1.0.27" or "3.0.0" version for creating configuration 
For all <"3.0.0" we use "1.0.27", for all which are after "3.0.0"  we use "3.0.0" 

so for "1.0.27" monitoring setting should be send together on connection creation
for "3.0.1" only connection settings are send

## Approach 2

- it uses the newest Kubernetes connection settings schema version, which is 3.0.0
- in case the request for the newest schema version fails (e.g. if the Operator connects to an older cluster version, which does not have the newest schema version yet) ---> fallback to the original request for schema version 1.0.27

## How can this be tested?

- Dy cluster should be up and running
- new k8s connection should be created with Automatic connection
- for "builtin:cloud.kubernetes"   version "3.0.0" msg 'These settings are overriding Environment settings' must NOT be shown on Monitoring settings

## Checklist
- [X] Unit tests have been updated/added
- [x] PR is labeled accordingly with a single label
- [X] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/dtclient/kubernetes_settings.go [hunk 1]
```diff
@@ -6,21 +6,30 @@ import (
...
+
+type monitoringSettings struct {
+	CloudApplicationPipelineEnabled bool `json:"cloudApplicationPipelineEnabled"`
+	OpenMetricsPipelineEnabled      bool `json:"openMetricsPipelineEnabled"`
+	EventProcessingActive           bool `json:"eventProcessingActive"`
+	EventProcessingV2Active         bool `json:"eventProcessingV2Active"`
+	FilterEvents                    bool `json:"filterEvents"`
+}
+
+func (d *postKubernetesSettings) UnmarshalJSON(b []byte) error {
```
Comment (chrismuellner) [inline src/dtclient/kubernetes_settings.go:31]: does this need to be its own method?

File: src/dtclient/kubernetes_settings.go [hunk 2]
```diff
@@ -67,33 +80,50 @@ type constraintViolations []struct {
...
-func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSystemUUID, scope string) (string, error) {
-	if kubeSystemUUID == "" {
-		return "", errors.New("no kube-system namespace UUID given")
+const (
+	schemaId                                    = "builtin:cloud.kubernetes"
+	defaultSchemaVersion                        = "1.0.27"
+	hierarchicalMonitoringSettingsSchemaVersion = "3.0.0"
+)
+
+func (dtc *dynatraceClient) GetSchemasVersion(schemaId string) (string, error) {
```
Comment (chrismuellner) [inline src/dtclient/kubernetes_settings.go:89]: I don't know if things changed, but didn't we want to skip the schema version check and just attempt to first post schema v3 and if it fails fall back to v1?

this would mean we have the same number of requests if the v3 schema is missing and only a single request if it's present/available.

**Review by chrismuellner — CHANGES_REQUESTED** [review]

File: src/dtclient/kubernetes_settings.go (hunk 1 above)
Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings.go]: Thanks for this point, I was able to remove it after renaming of field  monitoringSettings->MonitoringSettings
Actually the problem was in tests on 

var parsedBody []postKubernetesSettingsBody
err = json.Unmarshal(body, &parsedBody)

as monitoringSettings should be parsed as part of postKubernetesSettings, but I got an error so U added additional UnmarshalJSON. But after renaming monitoringSettings->MonitoringSettings it did work as exoected.

File: src/dtclient/kubernetes_settings.go (hunk 2 above)
Comment (0sewa0) [reply src/dtclient/kubernetes_settings.go:85]: is there a nice a reliable way to notice that its a fail because the schema v3 is not present, but otherwise everything is OK? 
I would prefer to avoid handling weird Status codes, and also always swallowing the error for the creation using v3 is not a nice thing aswell.

Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings.go]: I have change code according request of Christian and description in original issue, although to my opinion previous solution with version check was better. Please let me know in case we need to change it back.

File: src/dtclient/kubernetes_settings.go [hunk 3]
```diff
@@ -67,33 +72,13 @@ type constraintViolations []struct {
...
-		},
-	}
-
-	if scope != "" {
-		body[0].Scope = scope
-	}
+const (
+	schemaId                                    = "builtin:cloud.kubernetes"
+	defaultSchemaVersion                        = "1.0.27"
+	hierarchicalMonitoringSettingsSchemaVersion = "3.0.0"
```
Comment (chrismuellner) [inline src/dtclient/kubernetes_settings.go:78]: is this now the `defaultSchemaVersion`?

File: src/dtclient/kubernetes_settings.go [hunk 4]
```diff
@@ -133,6 +118,57 @@ func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSy
 	return resDataJson[0].ObjectId, nil
 }
 
+func (dtc *dynatraceClient) getKubernetesSettingBody(clusterLabel, kubeSystemUUID, scope string, schemaId string, isK8sHierarchicalMonitoringSettings bool) []postKubernetesSettingsBody {
```
Comment (chrismuellner) [inline src/dtclient/kubernetes_settings.go:121]: does not need to be a method

File: src/dtclient/kubernetes_settings.go [hunk 5]
```diff
@@ -133,6 +118,57 @@ func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSy
...
+			SchemaVersion: currentSchemaVersion,
+			Value: postKubernetesSettings{
+				Enabled:          true,
+				Label:            clusterLabel,
+				ClusterIdEnabled: true,
+				ClusterId:        kubeSystemUUID,
+			},
+		},
+	}
+	if scope != "" {
```
Comment (chrismuellner) [inline src/dtclient/kubernetes_settings.go:138]: can we always set this for the `body` as the default value is also an empty string?

File: src/dtclient/kubernetes_settings_test.go [hunk 6]
```diff
@@ -201,9 +204,29 @@ func TestDynatraceClient_GetSettingsForMonitoredEntities(t *testing.T) {
...
+		skipCert := SkipCertificateValidation(true)
+		dtc, err := NewClient(dynatraceServer.URL, apiToken, paasToken, skipCert)
+		require.NoError(t, err)
+		require.NotNil(t, dtc)
+
+		// act
+		actual, err := dtc.(*dynatraceClient).CreateOrUpdateKubernetesSetting(testName, testUID, testScope)
+
+		// assert
+		assert.NotNil(t, actual)
```
Comment (chrismuellner) [inline src/dtclient/kubernetes_settings_test.go:221]: use `require` to quick fail tests and not pollute output with follow up errors (see above, makes most sense for `NotNil`/`NoError` checks)

File: src/dtclient/kubernetes_settings.go (hunk 3 above)
Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings.go:78]: so we creating new setting with 2 schemas: "1.0.27" and  "3.0.0"
in case correct schema version > "3.0.0" but we try to use  "1.0.27" it still work and would not bring to error, so I suppose it is ok to have "1.0.27" as defaultSchemaVersion

File: src/dtclient/kubernetes_settings.go [hunk 7]
```diff
@@ -133,6 +118,57 @@ func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSy
...
+			OpenMetricsPipelineEnabled:      false,
+			EventProcessingActive:           false,
+			FilterEvents:                    false,
+			EventProcessingV2Active:         false,
+		}
+
+		body[0].Value.MonitoringSettings = &ms
+	}
+	return body
+}
```
Comment (0sewa0) [inline src/dtclient/kubernetes_settings.go:153]: I would split it this up a bit to be nicer:

We should only work with a list in the end so we don't do stuff like `body[0]...`

So would have a function for creating the base settings body: 
```
func createBaseKubernetesSettings(... ) postKubernetesSettingsBody {
	base := postKubernetesSettingsBody{
			SchemaId:      schemaId,
			Value: postKubernetesSettings{
				Enabled:          true,
				Label:            clusterLabel,
				ClusterIdEnabled: true,
				ClusterId:        kubeSystemUUID,
			},
		}
	if scope != "" {
		base.Scope = scope
	}
}
```

then have separate functions for the new Schema and the old Schema:
```
func createV3KubernetesSettingsBody(...) []postKubernetesSettingsBody {
	settings := createBaseKubernetesSettings(...)
	settings.Schema = hierarchicalMonitoringSettingsSchemaVersion
	ms := MonitoringSettings{
			CloudApplicationPipelineEnabled: true,
			OpenMetricsPipelineEnabled:      false,
			EventProcessingActive:           false,
			FilterEvents:                    false,
			EventProcessingV2Active:         false,
	}
	settings.Value.MonitoringSettings = &ms
	return []postKubernetesSettingsBody{settings}
}

func createV1KubernetesSettingsBody(...) []postKubernetesSettingsBody {
	settings := createBaseKubernetesSettings(...)
	settings.Schema = defaultSchemaVersion
	return []postKubernetesSettingsBody{settings}
}
```
by doing this you remove the weird ifs + bool logic, so you can make the code more readable and clear what it does

and the following reads a bit nicer:
```
func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSystemUUID, scope string) (string, error) {
	if kubeSystemUUID == "" {
		return "", errors.New("no kube-system namespace UUID given")
	}
	body := createV3KubernetesSettingsBody(clusterLabel, kubeSystemUUID, scope, schemaId)
	objectId, err := dtc.performCreateOrUpdateKubernetesSetting(body)
	if err != nil {
		if strings.Contains(err.Error(), strconv.Itoa(http.StatusNotFound)) {
			body = createV1KubernetesSettingsBody(clusterLabel, kubeSystemUUID, scope, schemaId)
			return dtc.performCreateOrUpdateKubernetesSetting(body)
		} else {
			return "", err
		}
	}
	return objectId, nil
}
```

File: src/dtclient/kubernetes_settings.go (hunk 4 above)
Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings.go:121]: for me it is question of taste, additionally I can test response body separately as json 
src/dtclient/kubernetes_settings_test.go:303
please let me know if we should really remove this method, thanks

File: src/dtclient/kubernetes_settings.go (hunk 3 above)
Comment (chrismuellner) [reply src/dtclient/kubernetes_settings.go:78]: we use v3 as default now and only fallback to v1 later though. We can also just rename the constant from `defaultSchemaVersion` to `schemaVersionV1`

File: src/dtclient/kubernetes_settings.go (hunk 4 above)
Comment (chrismuellner) [reply src/dtclient/kubernetes_settings.go:121]:
```suggestion
func getKubernetesSettingBody(clusterLabel, kubeSystemUUID, scope string, schemaId string, isK8sHierarchicalMonitoringSettings bool) []postKubernetesSettingsBody {
```
I mean to change it to a function, not completely remove it.

File: src/dtclient/kubernetes_settings_test.go (hunk 6 above)
Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings_test.go:221]: thanks for input, done

File: src/dtclient/kubernetes_settings.go (hunk 7 above)
Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings.go:153]: thanks, moved it to code

File: src/dtclient/kubernetes_settings.go (hunk 4 above)
Comment (annazaionchkovska) [reply src/dtclient/kubernetes_settings.go]: done

File: src/dtclient/kubernetes_settings.go [hunk 8]
```diff
@@ -133,6 +118,57 @@ func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSy
...
+
+func (dtc *dynatraceClient) CreateOrUpdateKubernetesSetting(clusterLabel, kubeSystemUUID, scope string) (string, error) {
+	if kubeSystemUUID == "" {
+		return "", errors.New("no kube-system namespace UUID given")
+	}
+	body := dtc.getKubernetesSettingBody(clusterLabel, kubeSystemUUID, scope, schemaId, true)
+	objectId, err := dtc.performCreateOrUpdateKubernetesSetting(body)
+	if err != nil {
+		if strings.Contains(err.Error(), strconv.Itoa(http.StatusNotFound)) {
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply src/dtclient/kubernetes_settings.go:163]: `k8serrors.IsNotFound` is meant to be used when trying to get a kubernetes object, and not when talking to the communicating with the dynatrace API.
(while both kubernetes and dynatrace API return the StatusCode 404, `k8serrors.IsNotFound` does some extra parsing that is kubernetes specific so I would avoid using it here)

TLDR:
let's leave it as it is now, its ain't perfect but using `k8serrors.IsNotFound` is not way here

**Review by chrismuellner — APPROVED** [review]
